service/topology_coordinator.cc

```diff
@@ -752,6 +753,87 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
         case global_topology_request::cleanup:
             co_await start_cleanup_on_dirty_nodes(std::move(guard), true);
             break;
+        case global_topology_request::keyspace_rf_change: {
+            while (true) {
+                auto tmptr = get_token_metadata_ptr();
+                sstring ks_name = *_topo_sm._topology.new_keyspace_rf_change_ks_name;
+                std::unordered_map<sstring, sstring> new_rf_per_dc = *_topo_sm._topology.new_keyspace_rf_change_rf_per_dc;
+                std::map<sstring, sstring> new_rf_per_dc_map{new_rf_per_dc.begin(), new_rf_per_dc.end()};
+                sstring req_uuid = *_topo_sm._topology.global_request_id;
+                std::vector<canonical_mutation> updates;
+                sstring error;
+                auto& ks = _db.find_keyspace(ks_name);
+                tablet_replica_calculation_result new_data{locator::tablet_map{8}, tablet_replica_calculation_status::success};
+
+                for (const auto& table : ks.metadata()->tables()) {
+                    auto repl_strategy = ks.get_replication_strategy_ptr()->maybe_as_tablet_aware();
+                    try {
+                        new_data = co_await service::calculate_tablet_replicas_for_new_rf(
+                                repl_strategy, table, tmptr, new_rf_per_dc_map);
+                    } catch(const std::exception& e) {
+                        new_data.status = tablet_replica_calculation_status::other;
+                        error = e.what();
+                        rtlogger.error("{}", error);
+                    }
+
+                    if (new_data.status != tablet_replica_calculation_status::success) {
+                        rtlogger.error("Couldn't process global_topology_request::keyspace_rf_change, status: {}, rf per dc map: {}",
+                                        new_data.status, new_data.new_tablet_map);
+                        updates.clear(); // remove all tablets mutations ...
+                        break;           // ... and only create mutations deleting the global req
+                    }
+
+                    auto tablet_id = new_data.new_tablet_map.first_tablet();
+                    while (true) {
+                        auto& tablet_info = new_data.new_tablet_map.get_tablet_info(tablet_id);
+                        auto last_token = new_data.new_tablet_map.get_last_token(tablet_id);
+                        updates.emplace_back(replica::tablet_mutation_builder(guard.write_timestamp(), table->id())
+                                                 .set_new_replicas(last_token, tablet_info.replicas)
+                                                 .set_stage(last_token, locator::tablet_transition_stage::allow_write_both_read_old)
+                                                 .set_transition(last_token, locator::tablet_transition_kind::rf_change)
+                                                 .build());
```

Comment:
I'm not sure how this would fly with 100K tablets generating 100K (canonical) mutations.
It would be better if we keep the tablet_mutation_builder around for batches of up to 1K tablets to reduce the per-mutation overhead and not create crazy long mutation vector.

Reply:
ping

Reply:
This has been discussed during tablets sync calls, and, to the best of my knowledge, the problem of too many mutations submitted at once is going to be addressed in the commitlog itself (https://github.com/scylladb/scylladb/issues/18161) rather than all the places that rely on it and may suffer from it. Also, we cannot split mutations in batches here, cause that'd result in loosing atomicity. 
That being said, to mitigate the problem, it's been proposed to merge/squash these mutations - but how? I see e.g. `squash_mutations` function, but only in tests, and `migration_manager::merge_schema_from`, which isn't probably what we want to use here, since it's not only the schema that we're modifying? @bhalevy ?

Reply:
you can apply rows to a single mutation partition (per table) one at a time.

Reply:
Could you point me to any source code example that implements that? Should I be looking at `mutation_partition.hh` and similar?

Reply:
I think you can simply instantiate here a `tablet_mutation_builder` per table, before the loop.
And repeatedly call:
```
                                                 .set_new_replicas(last_token, tablet_info.replicas)
                                                 .set_stage(last_token, locator::tablet_transition_stage::allow_write_both_read_old)
                                                 .set_transition(last_token, locator::tablet_transition_kind::rf_change)
```
per tablet.
(And don't forget to call co_await coroutine::maybe_yield, as there could be as many as 128K tablets to process.
Then, when done with all tablets, call:
```
updates.emplace_back(co_await make_canonical_mutation_gently(builder.build()));
```

Reply:
@bhalevy OK, thanks, I did that, but couldn't follow your instruction 1:1, specifically because there's no function called `make_canonical_mutation_gently` and I couldn't find anything similar, so I just ended up writing `updates.emplace_back(tablet_mutation_builder.build());` instead. But I can change it to something better.
Anyways, here's the diff to see what's exactly been changed:
```
diff --git a/service/topology_coordinator.cc b/service/topology_coordinator.cc
index a9a9179bf3..8366d3607d 100644
--- a/service/topology_coordinator.cc
+++ b/service/topology_coordinator.cc
@@ -781,21 +781,21 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                         break;           // ... and only create mutations deleting the global req
                     }
 
+                    replica::tablet_mutation_builder tablet_mutation_builder(guard.write_timestamp(), table->id());
                     auto tablet_id = new_tablet_map.first_tablet();
                     while (true) {
                         auto& tablet_info = new_tablet_map.get_tablet_info(tablet_id);
                         auto last_token = new_tablet_map.get_last_token(tablet_id);
-                        updates.emplace_back(replica::tablet_mutation_builder(guard.write_timestamp(), table->id())
-                                                 .set_new_replicas(last_token, tablet_info.replicas)
-                                                 .set_stage(last_token, locator::tablet_transition_stage::allow_write_both_read_old)
-                                                 .set_transition(last_token, locator::tablet_transition_kind::rf_change)
-                                                 .build());
-
+                        tablet_mutation_builder.set_new_replicas(last_token, tablet_info.replicas)
+                                               .set_stage(last_token, locator::tablet_transition_stage::allow_write_both_read_old)
+                                               .set_transition(last_token, locator::tablet_transition_kind::rf_change);
                         if (auto next_tablet = new_tablet_map.next_tablet(tablet_id); next_tablet.has_value()) {
                             tablet_id = *next_tablet;
                         } else {
+                            updates.emplace_back(tablet_mutation_builder.build());
                             break;
                         }
+                        co_await coroutine::maybe_yield();
                     }
                 }
```

Reply:
right, that's the direction...

Reply:
https://github.com/scylladb/scylladb/blob/2ce643d06bf04269661c1612b9b209a6ecc2a1b1/mutation/async_utils.hh#L40

Reply:
I had to rebase on top of the latest master to see it. Thanks, done as requested.

Reply:
Sorry I unresolved this conversation because I wanted to address this point.

> This has been discussed during tablets sync calls, and, to the best of my knowledge, the problem of too many mutations submitted at once is going to be addressed in the commitlog itself (https://github.com/scylladb/scylladb/issues/18161) rather than all the places that rely on it and may suffer from it.

This is an oversimplification and it's incorrect. If anyone during the tablets call made a promise that this is what's going to happen, then please back up from that promise because it's a false one.

Even if we manage to remove a limit inside commitlog, **there will still be a limit inside Raft**. The Raft log is **kept in memory**. There is a **limit on the size of the Raft log**, and therefore, there is a **limit on the size of a single Raft command**. We cannot assume that commands can be unbounded in size.

@tgrabiec @kostja @gleb-cloudius

Reply:
Sent this to mailing list, resolving conversation again, let's discuss on ML if you want